documentation/src/main/asciidoc/topics/ref_query_vector_search.adoc

```diff
@@ -0,0 +1,115 @@
+
+This is an example of kNN query:
+
+[source,sql]
+----
+from play.Item i where i.myVector <-> [7,7,7]~3
+----
+
+This query will find the items that have the `myVector` fields that are within `3` nearest neighbourhood from the vector `[7,7,7]`.
+
+Notice that in order to use this kind of search the entity, in our example `play.Item`, has to be `@Indexed` and
+the field, in our example `myVector`, should be annotated with `@Vector`.
+
+We support two kinds of vector field types:
+
+* `byte` / `Byte` (to work with byte vectors)
+* `float` / `Float` (to work with float vectors)
+
+You can have different vector fields on the same entity, but in any case you can have only one vector predicate on your queries.
+
+== Vector search parameters
+
+Both the k-value and the vector can be passed as query parameter.
+The k-value scalar can be expressed with the usual placeholder `:k` in the Ickle text.
+For the vector we can use either a placeholder for each term of the vector:
+
+[source,java]
+----
+Query<Item> query = cache.query("from play.Item i where i.floatVector <-> [:a,:b,:c]~:d");
+query.setParameter("a", 1);
+query.setParameter("b", 4.3);
+query.setParameter("c", 3.3);
+query.setParameter("d", 4);
+----
+
+Or a placeholder can be used for the entire vector:
+
+[source,java]
+----
+Query<Item> query = cache.query("from play.Item i where i.floatVector <-> [:a]~:b");
+query.setParameter("a", new float[]{7.1f, 7.0f, 3.1f});
+query.setParameter("b", 3);
+----
+
+== Vector field attributes
+
+It is always required to specify the **dimension** of the vector field.
+
+The other mapping attributes are optional, since Infinispan will have a default for each of them.
+You can configure them for instance to tune the desired accuracy / performances.
+
+=== Similarity
+
+Different `VectorSimilarity` algorithms are supported
+
+[%autowidth]
+|===
+| Value | Distance | Score | Note
+
+| `L2`
+|
+ifdef::backend-html5[stem:[d(x,y) = \sqrt{\sum_{i=1}^{n} (x_i - y_i)^2 } ]]
+ifdef::backend-pdf[`d(x,y) = sqrt(sum[i=1; i<n+1]( (x(i) - y(i))*(x(i) - y(i)) )`]
+|
+ifdef::backend-html5[stem:[s = \frac{1}{1+d^2}]]
+ifdef::backend-pdf[`s = 1/(1+d*d)`]
+| This is the {brandname} default
+| `INNER_PRODUCT`
+|
+ifdef::backend-html5[stem:[d(x,y) = \sum_{i=1}^{n} x_i \cdot y_i ]]
+ifdef::backend-pdf[`d(x,y) = sum[i=1; i< n+1] ( x(i)*y(i) ) `]
+|
+ifdef::backend-html5[stem:[s = \frac{1}{1+d}]]
+ifdef::backend-pdf[`s = 1/(1+d)`]
+| To use this similarity efficiently, both index and search vectors must be normalized
+| `MAX_INNER_PRODUCT`
+|
+ifdef::backend-html5[stem:[d(x,y) = \sum_{i=1}^{n} x_i \cdot y_i ]]
+ifdef::backend-pdf[`d(x,y) = sum[i=1; i<n+1] ( x(i)*y(i) )`]
+|
+ifdef::backend-html5[]
+s = \begin{cases}
+\frac{1}{1-d} & \text{if d < 0}\\
+d+1 & \text{otherwise}
+\end{cases}
+endif::[]
+ifdef::backend-pdf[`d<0 ? 1/(1-d) : d+1`]
+| This similarity does not require vector normalization
+| `COSINE`
```

Comment:
You probably can add a note that cosine similarity vectors cannot be of "zero magnitude", e.g. when a vector is all zeroes: `[0,0,0,... 0,0]`, the cosine is not defined and will result in an error.

Reply:
great idea! doing it